carma_ros2_utils/src/lifecycle_component_wrapper.cpp

```diff
@@ -0,0 +1,410 @@
+    std::bind(&LifecycleComponentWrapper::on_unload_node, this, _1, _2, _3));
+
+  listNodes_srv_ = create_service<ListNodes>(
+    "~/_container/list_nodes",
+    std::bind(&LifecycleComponentWrapper::on_list_nodes, this, _1, _2, _3));
+
+}
+
+////// END CARMA CHANGE /////
+
+LifecycleComponentWrapper::~LifecycleComponentWrapper()
+{
+  if (node_wrappers_.size()) {
+    RCLCPP_DEBUG(get_logger(), "Removing components from executor");
+    if (auto exec = executor_.lock()) {
+      for (auto & wrapper : node_wrappers_) {
+        exec->remove_node(wrapper.second.get_node_base_interface());
+      }
+    }
+  }
+}
+
+std::vector<LifecycleComponentWrapper::ComponentResource>
+LifecycleComponentWrapper::get_component_resources(
+  const std::string & package_name, const std::string & resource_index) const
+{
+  std::string content;
+  std::string base_path;
+  if (
+    !ament_index_cpp::get_resource(
+      resource_index, package_name, content, &base_path))
+  {
+    throw rclcpp_components::ComponentManagerException("Could not find requested resource in ament index");
+  }
+
+  std::vector<ComponentResource> resources;
+  std::vector<std::string> lines = rcpputils::split(content, '\n', true);
+  for (const auto & line : lines) {
+    std::vector<std::string> parts = rcpputils::split(line, ';');
+    if (parts.size() != 2) {
+      throw rclcpp_components::ComponentManagerException("Invalid resource entry");
+    }
+
+    std::string library_path = parts[1];
+    if (!rcpputils::fs::path(library_path).is_absolute()) {
+      library_path = base_path + "/" + library_path;
+    }
+    resources.push_back({parts[0], library_path});
+  }
+  return resources;
+}
+
+std::shared_ptr<rclcpp_components::NodeFactory>
+LifecycleComponentWrapper::create_component_factory(const ComponentResource & resource)
+{
+  std::string library_path = resource.second;
+  std::string class_name = resource.first;
+  std::string fq_class_name = "rclcpp_components::NodeFactoryTemplate<" + class_name + ">";
+
+  class_loader::ClassLoader * loader;
+  if (loaders_.find(library_path) == loaders_.end()) {
+    RCLCPP_INFO(get_logger(), "Load Library: %s", library_path.c_str());
+    try {
+      loaders_[library_path] = std::make_unique<class_loader::ClassLoader>(library_path);
+    } catch (const std::exception & ex) {
+      throw rclcpp_components::ComponentManagerException("Failed to load library: " + std::string(ex.what()));
+    } catch (...) {
+      throw rclcpp_components::ComponentManagerException("Failed to load library");
+    }
+  }
+  loader = loaders_[library_path].get();
+
+  auto classes = loader->getAvailableClasses<rclcpp_components::NodeFactory>();
+  for (const auto & clazz : classes) {
+    RCLCPP_INFO(get_logger(), "Found class: %s", clazz.c_str());
+    if (clazz == class_name || clazz == fq_class_name) {
+      RCLCPP_INFO(get_logger(), "Instantiate class: %s", clazz.c_str());
+      return loader->createInstance<rclcpp_components::NodeFactory>(clazz);
```

Comment:
Huh, I didn't even know this sort of class introspection/meta programming was possible in CPP.